test/e2e/clusterctl_upgrade.go

```diff
@@ -387,13 +394,34 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			input.PreWaitForCluster(managementClusterProxy, testNamespace.Name, workLoadClusterName)
 		}
 
+		// Build GroupVersionKind for Machine resources
+		machineListGVK := schema.GroupVersionKind{
+			Group: machineCRD.Spec.Group,
+			Kind:  machineCRD.Spec.Names.ListKind,
+		}
+
+		// Pick the first served version
+		for _, gvk := range machineCRD.Spec.Versions {
```

Comment:
```suggestion
		for _, version := range machineCRD.Spec.Versions {
```

Reply:
Thx, done and included in the single squashed commit